### PR TITLE
feat: Prevent collateral dust in liquidations

### PR DESCRIPTION
**Pull request by avniculae**
- prevent collateral dust in liquidations
- remove debtToCollateral interim variable in liquidations
- checks that collateral reserve was supplied by user in liquidation
- targeted tests for user account data
- scenario-based test for liquidations
- more fuzzing variations
- enables liquidation fee in gas tests for better gas reporting
- z3 prover script to make sure debtToLiquidate <= debtReserveBalance

prevents the accumulation of collateral dust in liquidations, which can pose a greater risk for bad debt accumulation by disincentivizing liquidations of remaining dust collateral that report deficit.

### CODE REVIEW

File: tests/mocks/MockSpoke.sol [hunk 1]
```diff
@@ -3,11 +3,26 @@
 pragma solidity ^0.8.0;
 
 import {Spoke, ISpoke, IHubBase, SafeCast, PositionStatusMap} from 'src/spoke/Spoke.sol';
+import {Test} from 'forge-std/Test.sol';
 
-contract MockSpoke is Spoke {
+/// @dev inherit from Test to exclude contract from forge size check
```
Comment (DhairyaSethi) [inline tests/mocks/MockSpoke.sol:8]: lol nice one, i think we can also just add 
bool public constant IS_TEST = true; 
bc i think that's the signal to forge that this is a test contract

File: tests/misc/debtToLiquidate.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+def mulDivUp(a, num, den):
+    return (a * num + den - 1) / den
+
+s = Solver()
+
+debtAssetPrice = Int('debtAssetPrice')
+s.add(1 <= debtAssetPrice, debtAssetPrice <= 10**30)
+debtAssetDecimals = Int('debtAssetDecimals')
+s.add(1 <= debtAssetDecimals, debtAssetDecimals <= 18)
+debtAssetUnit = Int('debtAssetUnit')
```
Comment (DhairyaSethi) [inline tests/misc/debtToLiquidate.py]:
```suggestion
debtAssetUnit = 10**debtAssetDecimals 
```
same with `collateralAssetUnit`, they dont need to be symbolic, two less vars to solve, nbd since this runs in time

File: src/spoke/libraries/LiquidationLogic.sol [hunk 3]
```diff
@@ -118,19 +119,23 @@ library LiquidationLogic {
     ISpoke.DynamicReserveConfig storage collateralDynConfig,
     LiquidateUserParams memory params
   ) external returns (bool) {
-    IHubBase collateralHub = collateralReserve.hub;
```
Comment (avniculae) [inline src/spoke/libraries/LiquidationLogic.sol:134]: removed this bc of stack too deep. I think it's fine to read it twice bc compiler is optimising it, wdyt @DhairyaSethi ?

File: src/spoke/Spoke.sol [hunk 4]
```diff
@@ -43,8 +43,7 @@ abstract contract Spoke is ISpoke, Multicall, NoncesKeyed, AccessManagedUpgradea
     LiquidationLogic.HEALTH_FACTOR_LIQUIDATION_THRESHOLD;
 
   /// @inheritdoc ISpoke
-  uint256 public constant DUST_DEBT_LIQUIDATION_THRESHOLD =
-    LiquidationLogic.DUST_DEBT_LIQUIDATION_THRESHOLD;
+  uint256 public constant DUST_LIQUIDATION_THRESHOLD = LiquidationLogic.DUST_LIQUIDATION_THRESHOLD;
```
Comment (yan-man) [inline src/spoke/Spoke.sol:50]: hm wondering if we need this in spoke?

File: src/spoke/libraries/LiquidationLogic.sol [hunk 5]
```diff
@@ -118,19 +119,23 @@ library LiquidationLogic {
...
+    uint256 collateralReserveBalance = collateralReserve.hub.previewRemoveByShares(
+      collateralReserve.assetId,
+      collateralPosition.suppliedShares
+    );
     _validateLiquidationCall(
```
Comment (yan-man) [inline src/spoke/libraries/LiquidationLogic.sol:139]: ah maybe it's a bigger change and not sure if it's worth it, but maybe we can split into 2 validations so that we can exit earlier. 1 set of validation is for non-HF related (ie flags, CF, balances), then validate HF afterwards, so that we can save a calcUserAccountData if the first validation fails

Comment (yan-man) [reply src/spoke/libraries/LiquidationLogic.sol:139]: something like this: https://github.com/aave/aave-v4/pull/859/files#diff-c2b8fc3739c2b0dd34b9513f5f01b4c4a225b524c9b34a4a1fa355105f994c47

ahh but maybe too late in the game and cost benefit too low given the breaking tests and edge case

File: src/spoke/libraries/LiquidationLogic.sol [hunk 6]
```diff
@@ -249,25 +252,46 @@ library LiquidationLogic {
...
+      DUST_LIQUIDATION_THRESHOLD;
+
+    if (
+      collateralToLiquidate > params.collateralReserveBalance ||
+      (leavesCollateralDust && debtToLiquidate < params.debtReserveBalance)
+    ) {
       collateralToLiquidate = params.collateralReserveBalance;
-      debtToCollateral = collateralToLiquidate.percentDivUp(liquidationBonus);
+
+      // decreases if collateralToLiquidate > params.collateralReserveBalance
```
Comment (yan-man) [inline src/spoke/libraries/LiquidationLogic.sol]:
```suggestion
      // debtToLiquidate is decreased if collateralToLiquidate > params.collateralReserveBalance
```

File: src/spoke/libraries/LiquidationLogic.sol [hunk 7]
```diff
@@ -249,25 +252,46 @@ library LiquidationLogic {
...
+      //  - if it decreases, it might bypass the debt dust condition
```
Comment (yan-man) [inline src/spoke/libraries/LiquidationLogic.sol]:
```suggestion
      //  - if so, debt dust could remain
```

File: src/spoke/libraries/LiquidationLogic.sol [hunk 8]
```diff
@@ -249,25 +252,46 @@ library LiquidationLogic {
...
+      // increases if leavesCollateralDust && debtToLiquidate < params.debtReserveBalance
```
Comment (yan-man) [inline src/spoke/libraries/LiquidationLogic.sol]:
```suggestion
      // debtToLiquidate is increased if (leavesCollateralDust && debtToLiquidate < params.debtReserveBalance), ensuring collateral reserve is fully liquidated
```

File: src/spoke/libraries/LiquidationLogic.sol [hunk 9]
```diff
@@ -249,25 +252,46 @@ library LiquidationLogic {
...
+      //  - if it increases, it will increase by at most DUST_LIQUIDATION_THRESHOLD (in base currency).
+      //    Since debtToLiquidate < params.debtReserveBalance and debt dust condition was enforced, it
+      //    is guaranteed that the increase will not make debtToLiquidate exceed params.debtReserveBalance
```
Comment (yan-man) [inline src/spoke/libraries/LiquidationLogic.sol]: `- can only increase by at most DUST_LIQUIDATION_THRESHOLD (in base currency)`
` Since debt dust condition was enforced, it
  is guaranteed that debtToLiquidate will never exceed params.debtReserveBalance`

File: src/spoke/libraries/LiquidationLogic.sol [hunk 10]
```diff
@@ -249,25 +252,46 @@ library LiquidationLogic {
...
+      //  - target health factor is ignored to prevent leaving dust
```
Comment (yan-man) [inline src/spoke/libraries/LiquidationLogic.sol]:
```suggestion
      //  - target health factor is bypassed to prevent leaving dust
```

File: src/spoke/libraries/LiquidationLogic.sol [hunk 11]
```diff
@@ -249,25 +252,46 @@ library LiquidationLogic {
...
       debtToLiquidate = collateralToLiquidate.mulDivUp(
         params.collateralAssetPrice * params.debtAssetUnit * PercentageMath.PERCENTAGE_FACTOR,
         params.debtAssetPrice * params.collateralAssetUnit * liquidationBonus
       );
     }
 
+    // make sure the liquidator intends to cover the necessary debt
+    // in case the amount was adjusted due to dust conditions
```
Comment (yan-man) [inline src/spoke/libraries/LiquidationLogic.sol]:
```
revert if the liquidator does not cover the necessary debt to prevent dust from remaining
```

File: src/spoke/libraries/LiquidationLogic.sol (hunk 10 above)
Comment (yan-man) [reply src/spoke/libraries/LiquidationLogic.sol]: or rm/edit below as it's the same comment

File: src/spoke/libraries/LiquidationLogic.sol (hunk 5 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol:139]: I agree, it would be great, but there are a few challenges including spoke size (which is exceeding max size in the PR you linked)

File: tests/misc/debtToLiquidate.py (hunk 2 above)
Comment (avniculae) [reply tests/misc/debtToLiquidate.py]: [ffabc76](https://github.com/aave/aave-v4/pull/796/commits/ffabc7672da3e9a23b0169f3f07b01d2811083fe)

File: src/spoke/Spoke.sol (hunk 4 above)
Comment (avniculae) [reply src/spoke/Spoke.sol:50]: I think it's good to expose it

File: src/spoke/libraries/LiquidationLogic.sol (hunk 11 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol]: [e0ffb03](https://github.com/aave/aave-v4/pull/796/commits/e0ffb032d96381aa72d859e3b1ebd6a38294ec46)

File: src/spoke/libraries/LiquidationLogic.sol (hunk 10 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol]: [e0ffb03](https://github.com/aave/aave-v4/pull/796/commits/e0ffb032d96381aa72d859e3b1ebd6a38294ec46)

File: src/spoke/libraries/LiquidationLogic.sol (hunk 9 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol]: [e0ffb03](https://github.com/aave/aave-v4/pull/796/commits/e0ffb032d96381aa72d859e3b1ebd6a38294ec46)

File: src/spoke/libraries/LiquidationLogic.sol (hunk 8 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol]: [e0ffb03](https://github.com/aave/aave-v4/pull/796/commits/e0ffb032d96381aa72d859e3b1ebd6a38294ec46)

File: src/spoke/libraries/LiquidationLogic.sol (hunk 7 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol]: [e0ffb03](https://github.com/aave/aave-v4/pull/796/commits/e0ffb032d96381aa72d859e3b1ebd6a38294ec46)

File: src/spoke/libraries/LiquidationLogic.sol (hunk 6 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol]: [e0ffb03](https://github.com/aave/aave-v4/pull/796/commits/e0ffb032d96381aa72d859e3b1ebd6a38294ec46)

File: src/spoke/Spoke.sol (hunk 4 above)
Comment (DhairyaSethi) [reply src/spoke/Spoke.sol:50]: yes definitely good to expose, it's still accessible from Spoke.getLiquidationLogic().DUST_THRESHOLD() but it's fine as is. if we have code size issues we can consider removing it here from since it's still accessible from elsewhere

File: src/spoke/interfaces/ISpoke.sol [hunk 12]
```diff
@@ -473,8 +476,9 @@ interface ISpoke is ISpokeBase, IMulticall, INoncesKeyed, IAccessManaged {
   function HEALTH_FACTOR_LIQUIDATION_THRESHOLD() external view returns (uint64);
 
   /// @notice Returns the minimum required remaining base currency amount after a partial liquidation.
-  /// @return The minimum debt amount considered as dust, denominated in USD with 26 decimals.
-  function DUST_DEBT_LIQUIDATION_THRESHOLD() external view returns (uint256);
+  /// @dev Applicable to both collateral and debt.
+  /// @return The minimum amount considered as dust, denominated in USD with 26 decimals.
```
Comment (yan-man) [inline src/spoke/interfaces/ISpoke.sol]: actually i think this is "max" amt considered as dust

File: src/spoke/libraries/LiquidationLogic.sol (hunk 5 above)
Comment (yan-man) [reply src/spoke/libraries/LiquidationLogic.sol:139]: yeah just an opt too, so not really necessary

File: src/spoke/interfaces/ISpoke.sol (hunk 12 above)
Comment (avniculae) [reply src/spoke/interfaces/ISpoke.sol]: [945b6a9](https://github.com/aave/aave-v4/pull/796/commits/945b6a9c66fcb4dae50facbbbb7998823ddac5b1)

File: tests/misc/debtToLiquidate.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+s.add(1 <= collateralAssetPrice, collateralAssetPrice <= 10**30)
+collateralAssetDecimals = Int('collateralAssetDecimals')
+s.add(1 <= collateralAssetDecimals, collateralAssetDecimals <= 18)
+collateralAssetUnit = 10**collateralAssetDecimals
+
+liquidationBonus = Int('liquidationBonus')
+s.add(PERCENTAGE_FACTOR <= liquidationBonus, liquidationBonus < PERCENTAGE_FACTOR * PERCENTAGE_FACTOR)
+
+debtReserveBalance = Int('debtReserveBalance')
+s.add(0 <= debtReserveBalance, debtReserveBalance <= 10**30)
```
Comment (yan-man) [inline tests/misc/debtToLiquidate.py:33]: should be strictly < debtReserveBalance?

File: tests/misc/debtToLiquidate.py [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+debtToLiquidate = Int('debtToLiquidate')
+s.add(0 <= debtToLiquidate, debtToLiquidate <= debtReserveBalance)
+
+collateralReserveBalance = Int('collateralReserveBalance')
+s.add(0 <= collateralReserveBalance, collateralReserveBalance <= 10**30)
```
Comment (yan-man) [inline tests/misc/debtToLiquidate.py:38]: strictly < ?

File: tests/misc/debtToLiquidate.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+debtAssetPrice = Int('debtAssetPrice')
+s.add(1 <= debtAssetPrice, debtAssetPrice <= 10**30)
+debtAssetDecimals = Int('debtAssetDecimals')
+s.add(1 <= debtAssetDecimals, debtAssetDecimals <= 18)
+debtAssetUnit = 10**debtAssetDecimals
+
+collateralAssetPrice = Int('collateralAssetPrice')
+s.add(1 <= collateralAssetPrice, collateralAssetPrice <= 10**30)
+collateralAssetDecimals = Int('collateralAssetDecimals')
+s.add(1 <= collateralAssetDecimals, collateralAssetDecimals <= 18)
```
Comment (yan-man) [inline tests/misc/debtToLiquidate.py:26]: nit: can tighten up decimals bounds now to 6? the min now in place

File: tests/misc/debtToLiquidate.py (hunk 13 above)
Comment (avniculae) [reply tests/misc/debtToLiquidate.py:33]: does not matter right?

File: tests/misc/debtToLiquidate.py (hunk 14 above)
Comment (avniculae) [reply tests/misc/debtToLiquidate.py:38]: same as above, does not matter right?

File: tests/misc/debtToLiquidate.py (hunk 15 above)
Comment (avniculae) [reply tests/misc/debtToLiquidate.py:26]: sure, we can, but there's no harm in having a bigger range here

File: src/spoke/interfaces/ISpoke.sol [hunk 16]
```diff
@@ -472,9 +475,10 @@ interface ISpoke is ISpokeBase, IMulticall, INoncesKeyed, IAccessManaged {
   /// @return The minimum health factor considered healthy, expressed in WAD (18 decimals) (e.g. 1e18 is 1.00).
   function HEALTH_FACTOR_LIQUIDATION_THRESHOLD() external view returns (uint64);
 
-  /// @notice Returns the minimum required remaining base currency amount after a partial liquidation.
-  /// @return The minimum debt amount considered as dust, denominated in USD with 26 decimals.
-  function DUST_DEBT_LIQUIDATION_THRESHOLD() external view returns (uint256);
+  /// @notice Returns the maximum amount considered as dust after a liquidation.
+  /// @dev Applicable to both collateral and debt.
```
Comment (miguelmtzinf) [inline src/spoke/interfaces/ISpoke.sol]:
```suggestion
  /// @notice Returns the maximum amount considered as dust for a user's collateral and debt balances after a liquidation.
```

File: src/spoke/libraries/LiquidationLogic.sol [hunk 17]
```diff
@@ -269,25 +272,42 @@ library LiquidationLogic {
...
+      collateralToLiquidate > params.collateralReserveBalance ||
+      (leavesCollateralDust && debtToLiquidate < params.debtReserveBalance)
+    ) {
       collateralToLiquidate = params.collateralReserveBalance;
-      debtToCollateral = collateralToLiquidate.percentDivUp(liquidationBonus);
+
+      // - debtToLiquidate is decreased if collateralToLiquidate > params.collateralReserveBalance (if so, debt dust could remain).
+      // - debtToLiquidate is increased if (leavesCollateralDust && debtToLiquidate < params.debtReserveBalance), ensuring collateral reserve
+      //   is fully liquidated (potentially bypassing the target health factor). Can only increase by at most DUST_LIQUIDATION_THRESHOLD (in
+      //   value terms). Since debt dust condition was enforced, it is guaranteed that debtToLiquidate will never exceed params.debtReserveBalance.
```
Comment (miguelmtzinf) [inline src/spoke/libraries/LiquidationLogic.sol]:
```suggestion
      // - `debtToLiquidate` is decreased if `collateralToLiquidate > params.collateralReserveBalance` (if so, debt dust could remain).
      // - `debtToLiquidate` is increased if `(leavesCollateralDust && debtToLiquidate < params.debtReserveBalance)`, ensuring collateral reserve
      //   is fully liquidated (potentially bypassing the target health factor). Can only increase by at most `DUST_LIQUIDATION_THRESHOLD` (in
      //   value terms). Since debt dust condition was enforced, it is guaranteed that `debtToLiquidate` will never exceed `params.debtReserveBalance`.
```

File: src/spoke/libraries/LiquidationLogic.sol [hunk 18]
```diff
@@ -269,25 +272,42 @@ library LiquidationLogic {
...
+      //   is fully liquidated (potentially bypassing the target health factor). Can only increase by at most `DUST_LIQUIDATION_THRESHOLD` (in
+      //   value terms). Since debt dust condition was enforced, it is guaranteed that `debtToLiquidate` will never exceed `params.debtReserveBalance`.
       debtToLiquidate = collateralToLiquidate.mulDivUp(
         params.collateralAssetPrice * params.debtAssetUnit * PercentageMath.PERCENTAGE_FACTOR,
         params.debtAssetPrice * params.collateralAssetUnit * liquidationBonus
       );
     }
 
+    // revert if the liquidator does not cover the necessary debt to prevent dust from remaining
+    require(params.debtToCover >= debtToLiquidate, ISpoke.MustNotLeaveDust());
```
Comment (miguelmtzinf) [inline src/spoke/libraries/LiquidationLogic.sol:308]: why is this always needed? in a simple liquidation (no dust in any side) we don't need to check this holds?

File: src/spoke/libraries/LiquidationLogic.sol [hunk 19]
```diff
@@ -347,14 +367,15 @@ library LiquidationLogic {
       maxDebtToLiquidate = debtToTarget;
     }
 
-    uint256 remainingDebtValue = (params.debtReserveBalance - maxDebtToLiquidate).mulDivDown(
-      params.debtAssetPrice.toWad(),
-      params.debtAssetUnit
-    );
+    bool leavesDebtDust = maxDebtToLiquidate < params.debtReserveBalance &&
+      (params.debtReserveBalance - maxDebtToLiquidate).mulDivDown(
```
Comment (miguelmtzinf) [inline src/spoke/libraries/LiquidationLogic.sol]: rounding up here and rounding down in `debtToLiquidate` re-calculation later would be safer, so rounding issues don't make the `debtToLiquidate` be greater that `debtReserveBalance`

File: src/spoke/libraries/LiquidationLogic.sol [hunk 20]
```diff
@@ -322,14 +345,15 @@ library LiquidationLogic {
       );
   }
 
-  /// @notice Calculates the maximum debt that can be liquidated.
-  /// @dev Dust debt less than the `DUST_DEBT_LIQUIDATION_THRESHOLD` cannot be left behind, unless the collateral reserve is fully liquidated.
-  function _calculateMaxDebtToLiquidate(
-    CalculateMaxDebtToLiquidateParams memory params
+  /// @notice Calculates the debt that should be liquidated.
+  /// @dev Generally, it returns the minimum of `debtToCover`, `debtReserveBalance` and `debtToTarget`.
+  /// If it would result in dust debt being left behind, the function returns `debtReserveBalance`.
```
Comment (miguelmtzinf) [inline src/spoke/libraries/LiquidationLogic.sol]:
```suggestion
  /// If dust debt would be left behind, it returns `debtReserveBalance` to ensure the debt is fully cleared and no dust is left.
```

File: src/spoke/libraries/LiquidationLogic.sol [hunk 21]
```diff
@@ -237,11 +239,15 @@ library LiquidationLogic {
       params.isUsingAsCollateral && params.collateralFactor > 0,
       ISpoke.CollateralCannotBeLiquidated()
     );
+    require(params.collateralReserveBalance > 0, ISpoke.ReserveNotSupplied());
     require(params.debtReserveBalance > 0, ISpoke.ReserveNotBorrowed());
   }
 
   /// @notice Calculates the liquidation amounts.
-  /// @dev Invoked by `liquidateUser` method.
+  /// @dev Invoked by `liquidateUser` method. To prevent accumulation of dust, it enforces one of the following conditions:
```
Comment (miguelmtzinf) [inline src/spoke/libraries/LiquidationLogic.sol]: I'm fine moving the comment about dust down, right before `_calculateDebtToLiquidate`

File: src/spoke/libraries/LiquidationLogic.sol (hunk 19 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol]: I agree it would fee safer bc we round down `debtToLiquidate`, but what I don't like about it is the fact that we would round `debtToLiquidate` differently is we enter the if branch or not. We want to round up the `debtToLiquidate` to help the user. Also, when assessing dust you want to round down to increase the chances of hitting the dust condition (which would also be different than collateral dust assessment rounding)

Personally I am fine with the current rounding given the confidence brought by the z3 prover

File: src/spoke/libraries/LiquidationLogic.sol (hunk 21 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol]: [d2e1836](https://github.com/aave/aave-v4/pull/796/commits/d2e18363539d5ba38aa18c0c942dca3c0c83bcf0)

File: src/spoke/libraries/LiquidationLogic.sol (hunk 18 above)
Comment (avniculae) [reply src/spoke/libraries/LiquidationLogic.sol:308]: yes, but at this point we don't know if debt to liquidate increased due to debt or collateral dust

Comment (StErMi) [reply src/spoke/libraries/LiquidationLogic.sol:308]: without this check it's possible (if we are in the dust debt scenario) that the liquidator would pay more than it is willing to pay (only possible if the allowance approved by the liquidator is `> params.debtToCover`)
